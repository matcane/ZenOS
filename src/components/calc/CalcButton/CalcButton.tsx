import FontAwesome6 from "@expo/vector-icons/FontAwesome6";
import { Pressable, TouchableOpacityProps } from "react-native";

import { ThemedText } from "@/components/core";
import { TButton } from "@/constants/calc";
import { useTheme } from "@/hooks/core";
import { baseStyle } from "@/styles/baseStyle";
import { coreStyles } from "@/styles/core";

const { container } = coreStyles;

type CalcButtonProps = TouchableOpacityProps & {
  button: TButton;
  isLast: boolean;
  itemWidth: number;
};

export default function CalcButton({ isLast, button, itemWidth, ...rest }: CalcButtonProps) {
  const theme = useTheme();

  const isNormal = isNaN(parseInt(button.char));
  const backgroundColor = isLast
    ? theme.primary
    : isNormal
      ? button.char === "."
        ? theme.container
        : theme.secondaryContainer
      : theme.container;
  const width = itemWidth * (isLast ? 2 : 1);

  const iconSize = baseStyle.fontMD.fontSize;

  const additionalStyles = {
    height: itemWidth - baseStyle.marginMD.margin * 2,
    width: width - baseStyle.marginMD.margin * 2,
    backgroundColor: backgroundColor,
  };

  return (
    <Pressable
      testID="calc-button"
      style={[
        container,
        baseStyle.justifyCenter,
        baseStyle.roundedFull,
        baseStyle.marginMD,
        additionalStyles,
      ]}
      {...rest}>
      {button.iconName ? (
        <FontAwesome6 testID="icon" name={button.iconName} size={iconSize} color={theme.text} />
      ) : (
        <ThemedText testID="text" style={baseStyle.fontMD}>
          {button.char}
        </ThemedText>
      )}
    </Pressable>
  );
}
