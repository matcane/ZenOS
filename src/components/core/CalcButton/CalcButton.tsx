import FontAwesome6 from "@expo/vector-icons/FontAwesome6";
import { TouchableOpacity, TouchableOpacityProps } from "react-native";

import { TButton } from "@/constants/calc";
import { useTheme } from "@/hooks/core";
import { baseStyle } from "@/styles/baseStyle";
import { coreStyles } from "@/styles/core";

import ThemedText from "../ThemedText/ThemedText";

const { fontMD, justifyCenter } = baseStyle;
const { container } = coreStyles;

type CalcButtonProps = TouchableOpacityProps & {
  button: TButton;
  isLast: boolean;
  index: number;
  itemWidth: number;
};

export default function CalcButton({ isLast, button, index, itemWidth, ...rest }: CalcButtonProps) {
  const theme = useTheme();

  const width = isLast ? itemWidth * 2 : itemWidth;
  const backgroundColor = isLast ? theme.primary : theme.container;

  const iconSize = fontMD;

  const additionalStyles = {
    height: itemWidth,
    width: width,
    backgroundColor: backgroundColor,
  };

  return (
    <TouchableOpacity style={[container, justifyCenter, additionalStyles]} {...rest}>
      {button.iconName ? (
        <FontAwesome6 name={button.iconName} size={iconSize} color={theme.text} />
      ) : (
        <ThemedText style={fontMD}>{button.char}</ThemedText>
      )}
    </TouchableOpacity>
  );
}
