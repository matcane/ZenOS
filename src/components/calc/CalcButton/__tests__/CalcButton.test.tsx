import { render, fireEvent } from "@testing-library/react-native";

import CalcButton from "../CalcButton";

const ITEM_WIDTH = 50;

const button = {
  iconName: "plus",
  char: "+",
  action: "",
};

const props = {
  button,
  isLast: false,
  itemWidth: ITEM_WIDTH,
};

jest.mock("@react-native-firebase/auth", () => {
  return {
    onAuthStateChanged: jest.fn(),
    signInWithEmailAndPassword: jest.fn(),
    createUserWithEmailAndPassword: jest.fn(),
    signOut: jest.fn(),
  };
});

jest.mock("@react-native-firebase/firestore", () => {
  return {
    collection: jest.fn(() => ({
      doc: jest.fn(() => ({
        get: jest.fn(() => Promise.resolve({ exists: true, data: () => ({}) })),
        set: jest.fn(() => Promise.resolve()),
        update: jest.fn(() => Promise.resolve()),
        delete: jest.fn(() => Promise.resolve()),
      })),
    })),
  };
});

describe("CalcButton component", () => {
  it("renders correctly", () => {
    const tree = render(<CalcButton {...props} />);

    expect(tree).toMatchSnapshot();
  });

  it("renders icon when iconName is provided", () => {
    const { getByTestId } = render(<CalcButton {...props} />);

    const icon = getByTestId("icon");
    expect(icon).toBeTruthy();
  });

  it("renders text when iconName is not provided", () => {
    const newButton = {
      iconName: "",
      char: "+",
      action: "",
    };

    const { getByTestId } = render(<CalcButton {...props} button={newButton} />);

    const text = getByTestId("text");
    expect(text).toBeTruthy();
    expect(text.props.children).toBe("+");
  });

  it("has correct styles when isLast is true", () => {
    const { getByTestId } = render(<CalcButton {...props} isLast={true} />);

    const buttonContainer = getByTestId("calc-button");
    const style = buttonContainer.props.style;
    expect(style).toEqual([
      { alignItems: "center", padding: 20 },
      { justifyContent: "center" },
      { borderRadius: 999 },
      { margin: 10 },
      { backgroundColor: "#2196f3", height: 30, width: 80 },
    ]);
  });

  it("has correct styles when isLast is false", () => {
    const { getByTestId } = render(<CalcButton {...props} isLast={false} />);

    const buttonContainer = getByTestId("calc-button");
    const style = buttonContainer.props.style;
    expect(style).toEqual([
      { alignItems: "center", padding: 20 },
      { justifyContent: "center" },
      { borderRadius: 999 },
      { margin: 10 },
      { backgroundColor: "#a0a0a0", height: 30, width: 30 },
    ]);
  });

  it("calls onPress when pressed", () => {
    const fn = jest.fn();
    const { getByTestId } = render(<CalcButton {...props} onPress={fn} />);

    const buttonContainer = getByTestId("calc-button");
    fireEvent.press(buttonContainer);
    expect(fn).toHaveBeenCalledTimes(1);
  });
});
