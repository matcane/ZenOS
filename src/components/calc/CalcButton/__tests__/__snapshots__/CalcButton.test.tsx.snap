// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalcButton component renders correctly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "alignItems": "center",
        "padding": 20,
      },
      {
        "justifyContent": "center",
      },
      {
        "borderRadius": 999,
      },
      {
        "margin": 10,
      },
      {
        "backgroundColor": "#a0a0a0",
        "height": 30,
        "width": 30,
      },
    ]
  }
  testID="calc-button"
>
  <Text />
</View>
`;
